src/runtime/kernel-signal.jl

```diff
@@ -0,0 +1,90 @@
+struct KernelException <: Exception
+    device::ROCDevice
+    exstr::Union{String, Nothing}
+end
+
+function Base.showerror(io::IO, err::KernelException)
+    print(io, "KernelException: exception(s) thrown during kernel execution on device $(err.device)")
+    if err.exstr !== nothing
+        println(io, ":")
+        print(io, err.exstr)
+    end
+end
+
+mutable struct ROCKernelSignal
+    signal::ROCSignal
+    queue::ROCQueue
+    kernel::ROCKernel
+    exception::Union{Exception, Nothing}
+    @atomic active::Bool
+
+    function ROCKernelSignal(
+        signal::ROCSignal, queue::ROCQueue, kernel::ROCKernel;
+    )
+        kersig = new(signal, queue, kernel, nothing, true)
+        finalizer(kersig) do k
+            cleanup!(k; finished=true)
+        end
+        kersig
+    end
+end
+
+const _active_kernels = WeakKeyDict{ROCQueue, Vector{ROCKernelSignal}}()
+
+function Base.wait(
+    kersig::ROCKernelSignal; check_exceptions::Bool = true, signal_kwargs...,
+)
+    finished = try
+        wait(kersig.signal; signal_kwargs...)
+        true
+    catch err
+        if isa(err, SignalTimeoutException) && SIGNAL_TIMEOUT_KILL_QUEUE[]
+            kill_queue!(kersig.queue)
+        end
+        isnothing(kersig.exception) && (kersig.exception = err;)
+        false
+    finally
+        queue_error = kersig.queue.status != HSA.STATUS_SUCCESS
+        # If it is QueueError, `kill_queue!` will fill `exception` field.
+        queue_error && kill_queue!(kersig.queue)
+    end
+
+    cleanup!(kersig; finished, check_exceptions)
+    finished
+end
+
+function cleanup!(
+    kersig::ROCKernelSignal; finished::Bool, check_exceptions::Bool = true,
+)
+    (@atomic kersig.active) || return nothing
+    @atomic kersig.active = false
+
+    unpreserve!(kersig)
+    finalize(kersig.kernel)
```

Comment:
```suggestion
```
We don't know that all references are gone at this point (and in fact, they probably aren't, since this function inherently preserves the object until the end of its execution).